modelmapper/stats.py

```diff
@@ -0,0 +1,474 @@
+import decimal
+import datetime
+from collections import Counter
+from typing import Any, NamedTuple
+
+from modelmapper.types import (
+    HasBoolean,
+    HasDateTime,
+    HasDecimal,
+    HasDollar,
+    HasInt,
+    HasNull,
+    HasPercent,
+    HasString,
+)
+
+
+class FieldStats(NamedTuple):
+    counter: Any
+    max_int: 'FieldStats' = 0
+    max_pre_decimal: 'FieldStats' = 0
+    max_decimal_scale: 'FieldStats' = 0
+    max_string_len: 'FieldStats' = 0
+    datetime_formats: 'FieldStats' = None
+    len: 'FieldStats' = 0
+
+
+class UserInferenceRequired(ValueError):
+    """Thrown when we need the prompt the user for more information"""
+    def __init__(self, value_type, message):
+        super().__init__(message)
+        self.value_type = value_type
+
+
+class NoMatchFound(ValueError):
+    """Thrown when we can't match a datatype"""
+    pass
+
+
+class InconsistentData(ValueError):
+    """Thrown when we detect inconsistent data within a field"""
+    pass
+
+
+class TypeMatcher(object):
+    """
+    Matches a value to a type
+    """
+    def match(self, item, settings):
+        return False
+
+    def is_exclusive(self):
+        """
+        Return True if a match of this type should be the _only_ reported type for a field
+        """
+        return False
+
+
+class TypeAccumulator(object):
+    """
+    Accumulates some value over every item it sees
+    """
+    def reset(self):
+        """
+        Resets the accumulator
+        """
+        pass
+
+    def inspect(self, field_name, item, settings):
+        """
+        Inspects the item
+        """
+        return False
+
+    def collect(self):
+        """
+        Sets the value of a field on the FieldStats object
+        """
+        pass
+
+
+class InSettingsMatcher(TypeMatcher):
+    """
+    Abstract Matacher -- This should be subclassed and not created directly.
+
+    Matches a value by comparing it to values in
+    the settings object.
+
+    >>> settings.test = ['1', 'y', 'true']
+    >>> InSettingsMatcher('test').match('test', settings)
+    False
+    >>> settings.test = ['1', 'y', 'true']
+    >>> InSettingsMatcher('test').match('1', settings)
+    True
+    """
+    def __init__(self, field):
```

Comment:
I think the interface maybe should be that is takes settings in `__init__`? 
Example: `InSettingsMatcher(settings, 'test').match('1')`

Reply:
I'm 50/50 on it. I didn't like binding the matchers to the settings because we modify the settings in some cases (unfamiliar date-format, etc.) and I don't like that a change outside the class has the ability to modify the behavior of the class itself.

Reply:
Similarly, I didn't like passing in the settings object to the StatsCollector, I think the proper behavior would be to provide the set of matchers from the mapper, or have a helper method that builds the matchers from the settings object.